R/matching.R

```diff
@@ -0,0 +1,140 @@
+#' same behaviour as [`match`]. Use `Inf` (default) to match to the closest
+#' value without restrictions. Could be of length one or the same length as
+#' `table`.
+#' @param duplicates `character(1)`, it is possible that many `x` match one
+#' element in `table`. For `duplicates="keep"` (default) all of them are
+#' returned. For `duplicates="closest"` just the `x` that is closest to the
+#' element in `table` gets an index (all the other will get `nomatch`). If the
+#' difference is exactly identical the first one in `x` will get an index. For
+#' `duplicates="remove"` all duplicated matches are returned as `nomatch`. If
+#' a single element in `x` matches to multiple elements in `table` the *closest*
+#' is returned for `duplicates="keep"` or `duplicates="duplicates"` (*keeping*
+#' multiple matches isn't possible in this case because the implentation relies
+#' on [`findInterval`]). If the difference is identical the index of the
+#' smaller element in `table` is returned.
+#' For `duplicates="remove"` all multiple matches are returned as `nomatch`.
+#' @param nomatch `numeric(1)`, if the difference
+#' between the value in `x` and `table` is larger than
+#' `tolerance` `nomatch` is returned.
+#'
+#' @return `closest` returns an `integer` vector of the same length as `x`
+#' giving the closest position in `table` of the first match or `nomatch` if
+#' there is no match.
+#'
+#' @rdname matching
+#' @author Sebastian Gibb
+#' @seealso [`match`]
+#' @aliases closest
+#' @export
+#' @examples
+#' ## Define two vectors to match
+#' x <- c(1, 3, 5)
+#' y <- 1:10
+#'
+#' ## Compare match and closest
+#' match(x, y)
+#' closest(x, y)
+#'
+#' ## If there is no exact match
+#' x <- x + 0.1
+#' match(x, y) # no match
+#' closest(x, y)
+#'
+#' ## Some new values
+#' x <- c(1.11, 45.02, 556.45)
+#' y <- c(3.01, 34.12, 45.021, 46.1, 556.449)
+#'
+#' ## Using a single tolerance value
+#' closest(x, y, tolerance = 0.01)
+#'
+#' ## Using a value-specific tolerance accepting differences of 20 ppm
+#' closest(x, y, tolerance = x * ppm(20))
+#'
+#' ## Same using 50 ppm
+#' closest(x, y, tolerance = x * ppm(50))
+#'
+#' ## Sometimes multiple elements in `x` match to `table`
+#' x <- c(1.6, 1.75, 1.8)
+#' y <- 1:2
+#' closest(x, y, tolerance = 0.5)
+#' closest(x, y, tolerance = 0.5, duplicates = "closest")
+#' closest(x, y, tolerance = 0.5, duplicates = "remove")
+closest <- function(x, table, tolerance = Inf,
+                    duplicates = c("keep", "closest", "remove"),
+                    nomatch = NA_integer_) {
+
+    if (!is.numeric(tolerance) || tolerance < 0)
+        stop("'tolerance' has to be a 'numeric' larger or equal zero.")
+
+    if (!is.numeric(nomatch) || length(nomatch) != 1L)
+        stop("'nomatch' has to be a 'numeric' of length one.")
+
+    tolerance <- rep_len(tolerance, length(table)) + sqrt(.Machine$double.eps)
+    duplicates <- match.arg(duplicates)
+    ntable <- length(table)
```

Comment:
You could move this 2 lines up and then have 
`tolerance <- rep_len(tolerance, ntable) + sqrt(.Machine$double.eps)`, saves you one `length` call.

Reply:
Thanks! Changed it.